### ISSUE TITLE
Peer cluster nodes

### ISSUE THREAD

**Issue by maxim-ge** [voedger/voedger#1891]
- Numbers
- Optimize, engines per processor: `DeployApp(name istructs.AppQName, def appdef.IAppDef, engines [cluster.ProcessorKind_Count]int)`
- ResellerAdmin:  uses different workspaces: Query: CreateunTillPayment  
- BootstrapLeader

**Comment by maxim-ge** [comment]
### Context and Need for `host.docker.internal`

When Docker containers are created, they are often isolated from the host machine's network for security and stability. This means that, by default, a container doesn't have direct access to services running on the host (e.g., a database, API server, or any local development server you might be running). This can be problematic if the application inside the container needs to communicate with something on the host.

For example:
- If you’re running a database on the host machine and want your containerized application to connect to it, the container needs a way to "see" the host.
- Similarly, if you’re developing and want to connect your Dockerized application to a development server or API that’s only accessible on your host, the container needs to be able to reach it.

### Solution: `host.docker.internal`

`host.docker.internal` is a special hostname provided by Docker to refer to the host machine from within a container. It acts as a shortcut, resolving to the IP address of the Docker host. However, this hostname is only natively supported on macOS and Windows versions of Docker, not on Linux.

### `host-gateway` Option

On Linux, or for cross-platform compatibility, Docker introduced a special option called `host-gateway`. This option allows the container to resolve the host IP address dynamically, providing a reliable way for containers to access the host, regardless of platform.

Here’s how it works in detail:

- **`--add-host=host.docker.internal:host-gateway`**: This command option tells Docker to create an entry in the container’s `/etc/hosts` file that maps the name `host.docker.internal` to the special gateway IP address of the host.
  
- When you add `--add-host=host.docker.internal:host-gateway`, Docker resolves `host-gateway` to the host's gateway IP (the IP address that the container network uses to reach the host). This effectively means:
  
  ```
  host.docker.internal -> Host's gateway IP address
  ```

### Why Use `host.docker.internal`?

1. **Cross-Platform Compatibility**:
   - Using `host.docker.internal:host-gateway` makes the hostname `host.docker.internal` available across all operating systems. This ensures that your Docker setup can work on Linux, macOS, and Windows in a consistent way.
   
2. **Ease of Access**:
   - By setting up `host.docker.internal`, you can use this hostname within your container’s application configuration, environment variables, or code to refer to the host. This avoids hardcoding an IP address, which can change depending on the Docker network setup.

3. **Dynamic and Reliable**:
   - Since Docker dynamically resolves the host's gateway IP using `host-gateway`, it is less brittle than hardcoding an IP address. The gateway IP might vary across Docker networks or restarts, so this option provides a stable reference.

### Example Use Cases

Imagine you have a web application running in a Docker container that needs to connect to a database running on your host machine. Without `host.docker.internal`, you’d have to determine the host IP and hardcode it, or use another complex networking setup. With `host.docker.internal`:

- You could configure the application to connect to `host.docker.internal` as the database host, simplifying development and ensuring the connection is robust across platforms.
  
### Example Command Inside the Container

Suppose you have the following database configuration in your application:

```json
{
  "database_host": "host.docker.internal",
  "database_port": 5432
}
```

Inside the container, `host.docker.internal` will resolve to the host's gateway IP, allowing the containerized application to communicate directly with the database on the host.

### Summary

In summary, `--add-host=host.docker.internal:host-gateway` is a way to add a hostname entry for `host.docker.internal` in the container that maps to the host’s IP address. This allows applications within the container to communicate with services on the Docker host using a simple, consistent hostname. This feature enhances portability, cross-platform compatibility, and simplifies network configuration for containers needing access to host services.

**Comment by maxim-ge** [comment]
[Dynamic Nginx Load Balancing](https://chatgpt.com/share/6738816e-3bc8-800b-b294-820e43994ae9)

**Comment by maxim-ge** [comment]
## Routing using division reminder

If you want to route requests based on the **remainder of `wsid` divided by the number of servers (`NumServers`)**, you can achieve this using NGINX's `map` directive and modulo arithmetic with the `split_clients` module. Here's how you can configure it:

### Configuration:

```nginx
http {
    # Number of servers
    map $uri $wsid {
        default "";
        ~^/api/[^/]+/[^/]+/([^/]+)/ { 
            $1;  # Extract the wsid
        }
    }

    map $wsid $remainder {
        default 0;
        ~^\d+$ $wsid;
    }

    # Use split_clients for modulo-based routing
    split_clients "${remainder}" $target_server {
        0%6 vvm1;
        1%6 vvm; 5th

Here is the complete configuration to use the remainder (`wsid % NumServers`) for routing:

### Configuration:

```nginx
http {
    # Define number of servers
    upstream backend {
        server vvm1;
        server vvm2;
        server vvm3;
        server vvm4;
        server vvm5;
        server vvm6;
    }

    # Extract the wsid from the URL
    map $uri $wsid {
        default "";
        ~^/api/[^/]+/[^/]+/([^/]+)/ $1;  # Capture the wsid part
    }

    # Compute the remainder: wsid % NumServers
    map $wsid $server_index {
        default 0;
        ~^\d+$ $wsid % 6;  # Use modulo operator with 6 servers
    }

    # Map the remainder to a specific server
    map $server_index $target_server {
        0 vvm1;
        1 vvm2;
        2 vvm3;
        3 vvm4;
        4 vvm5;
        5 vvm6;
    }

    server {
        listen 80;

        location / {
            # Proxy requests based on the calculated target server
            set $backend "http://$target_server";
            proxy_pass $backend;
        }
    }
}
```

### Explanation:

1. **`map $uri $wsid`:**
   - Extracts the `wsid` from the URL using a regex. For example, from `/api/owner/name/12345/q.funcName`, it extracts `12345`.

2. **`map $wsid $server_index`:**
   - Computes the modulo operation (`wsid % NumServers`), where `NumServers` is 6 in this case.
   - The `~^\d+$` ensures `wsid` is a numeric value. If it’s not numeric, it defaults to `0`.

3. **`map $server_index $target_server`:**
   - Maps the remainder (0-5) to the corresponding server (`vvm1` to `vvm6`).

4. **`proxy_pass $backend`:**
   - Proxies the request to the specific server determined by the modulo calculation.

### Steps to Enable:
1. Save this configuration to your NGINX configuration file.
2. Test the configuration:
   ```bash
   nginx -t
   ```
3. Reload NGINX:
   ```bash
   nginx -s reload
   ```

### Example Workflow:

- URL: `/api/owner/name/12345/q.funcName`
  - Extracted `wsid`: `12345`
  - `wsid % 6`: `3`
  - Routed to `vvm4`.

This method ensures predictable routing based on the `wsid` modulo operation, distributing requests evenly across your servers.

**Comment by maxim-ge** [comment]
## Routing using CRC32 

To route based on the CRC32 hash of the `wsid` extracted from the URL path, you can use NGINX's `map` directive in combination with a regex to extract the `wsid`. The `hash` directive in the `upstream` block will then use this value for consistent hashing.

Here's how you can configure it:

### Configuration:

```nginx
http {
    # Define the upstream group
    upstream backend {
        hash $wsid consistent;
        server vvm1;
        server vvm2;
        server vvm3;
        server vvm4;
        server vvm5;
        server vvm6;
    }

    # Extract wsid using a map and regex
    map $uri $wsid {
        default "";
        ~^/api/[^/]+/[^/]+/([^/]+)/ { 
            # The capturing group extracts <wsid>
            $1;
        }
    }

    server {
        listen 80;

        location / {
            # Proxy the request to the upstream
            proxy_pass http://backend;
        }
    }
}
```

### Explanation:

1. **`map $uri $wsid`:**
   - Uses a regular expression to extract the `wsid` from the URL path.
   - The regex `~^/api/[^/]+/[^/]+/([^/]+)/`:
     - Matches `/api/<AppQName.owner>/<AppQName.name>/<wsid>/`.
     - Captures the `<wsid>` part using `([^/]+)`.

2. **`hash $wsid consistent;`:**
   - Routes requests to the upstream servers based on the CRC32 hash of the extracted `wsid`.
   - The `consistent` keyword ensures minimal redistribution of traffic if servers are added or removed.

3. **`proxy_pass http://backend;`:**
   - Sends the request to the appropriate backend server determined by the hash of `$wsid`.

4. **Fallback Behavior:**
   - If the `wsid` cannot be extracted (e.g., malformed URL), `$wsid` will default to an empty string. You may want to handle this with a fallback upstream or a custom error message.

### Steps to Enable:
1. Save the configuration to your NGINX configuration file (e.g., `/etc/nginx/nginx.conf`).
2. Test the configuration:
   ```bash
   nginx -t
   ```
3. Reload NGINX:
   ```bash
   nginx -s reload
   ```

This setup ensures that requests are distributed across `vvm1-vvm6` based on the hash of the `wsid` extracted from the URL.
